gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java

```diff
@@ -0,0 +1,106 @@
+ */
+public class Crash2ExceptionHandler implements Thread.UncaughtExceptionHandler {
+  private static final Logger LOG = Logger.getLogger(Crash2ExceptionHandler.class.getName());
+
+  // TODO(baldwinn860): Send to production url when we get approval.
+  private static final String CRASH_REPORT_URL_BASE = "https://clients2.google.com/cr/staging_report?";
+  private static final String CRASH_REPORT_PRODUCT = "GAPID";
+  private static final String CRASH_REPORT_VERSION = "Client:" + GAPID_VERSION.toString();
+  private static final String CRASH_REPORT_VERSION_ENCODED;
+  private static final String CRASH_REPORT_URL_FIELDS;
+  private static final String CRASH_REPORT_URL;
+  static { 
+    String temp;
+    try {
+      temp = URLEncoder.encode(CRASH_REPORT_VERSION, "UTF-8");
+    } catch(UnsupportedEncodingException e) {
+      temp = "UnknownVersion";
+    }
+    CRASH_REPORT_VERSION_ENCODED = temp;
+    CRASH_REPORT_URL_FIELDS = "product=" + CRASH_REPORT_PRODUCT + "&version=" + CRASH_REPORT_VERSION_ENCODED;
+    CRASH_REPORT_URL = CRASH_REPORT_URL_BASE + CRASH_REPORT_URL_FIELDS;
+  }
+
+  private final Thread.UncaughtExceptionHandler previousHandler;
+
+  public static void Register() {
+    if (!(Thread.getDefaultUncaughtExceptionHandler() instanceof Crash2ExceptionHandler)) {
+      Crash2ExceptionHandler handler = new Crash2ExceptionHandler(Thread.getDefaultUncaughtExceptionHandler());
+      Thread.setDefaultUncaughtExceptionHandler(handler);
+    }
+  }
+
+  public Crash2ExceptionHandler(Thread.UncaughtExceptionHandler oldHandler) {
+    previousHandler = oldHandler;
+  }
+
+  @Override
+  public void uncaughtException(Thread thread, Throwable thrown) {
+    Scheduler.EXECUTOR.submit(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          // Creates a connection to crash2
+          OkHttpClient client = new OkHttpClient();
+          Response response = client.newCall(new Request.Builder()
+                  .url(CRASH_REPORT_URL)
+                  .post(new MultipartBuilder()
+                      .type(MultipartBuilder.FORM)
+                      .addFormDataPart("product", CRASH_REPORT_PRODUCT)
+                      .addFormDataPart("version", CRASH_REPORT_VERSION)
+                      .addFormDataPart("exception_info", Throwables.getStackTraceAsString(thrown))
+                      .build())
+                  .build())
+              .execute();
+          if (response.isSuccessful()) {
+            LOG.log(Level.INFO, "Crash Report Uploaded Successfully; Crash Report ID: " + response.body().string());
+          } else {
+            LOG.log(Level.SEVERE, "Crash Report Not Uploaded; Response Code: " + response.code());
+          }
+        } catch (IOException e) {
+          LOG.log(Level.SEVERE, "Unable to upload exception to crash2", e);
+        } finally {
+          // Pass the exception back to the os to get logged
+          previousHandler.uncaughtException(thread, thrown);
```

Comment:
Since this is now happening on a different thread, it's probably better to just invoke it after scheduling the crash2 stuff.